File: crates/traderview-db/src/brrrr.rs
//! BRRRR — Buy, Rehab, Rent, Refinance, Repeat.
//!
//! The real-estate capital-recycling play: buy a distressed property
//! cheap, rehab it to force appreciation, rent it, then refinance against
//! the higher after-repair value (ARV) and pull your cash back out to do
//! it again. The deal succeeds when the cash-out refinance recovers all
//! the cash you put in — what's left in the deal still earns rent, so the
//! cash-on-cash return goes infinite.
//!
//! This nets the refinance proceeds against the total cash invested to
//! report the cash left in the deal, the post-refi mortgage and cash
//! flow, the equity captured by the rehab, and the cash-on-cash return.
//!
//! Pure compute — the refinance P&I uses the shared
//! [`crate::mortgage_amortization::monthly_payment`].

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct BrrrrInput {
    pub purchase_price_usd: f64,
    pub rehab_cost_usd: f64,
    /// Acquisition closing costs (title, transfer, points on the buy).
    pub purchase_closing_usd: f64,
    /// After-repair value — the appraised value the refinance is based on.
    pub after_repair_value_usd: f64,
    /// Refinance loan-to-value, e.g. 75 for a 75% cash-out refi.
    pub refi_ltv_pct: f64,
    pub refi_apr_pct: f64,
    pub refi_term_months: u32,
    /// Refinance closing costs, netted out of the cash pulled.
    pub refi_closing_usd: f64,
    pub monthly_rent_usd: f64,
    /// All-in monthly operating cost EXCLUDING the mortgage (tax,
    /// insurance, maintenance, vacancy, management).
    pub monthly_operating_usd: f64,
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct BrrrrResult {
    /// Purchase + rehab + acquisition closing — the cash that goes in.
    pub total_cash_invested_usd: f64,
    /// ARV × refi LTV.
    pub refi_loan_usd: f64,
    /// Loan proceeds net of refinance closing costs — cash back to you.
    pub cash_out_usd: f64,
    /// Cash invested minus cash pulled. Negative = you pulled out more
    /// than you put in (and still own the asset).
    pub cash_left_in_deal_usd: f64,
    /// True when the refinance recovered all the invested cash.
    pub all_cash_recovered: bool,
    pub monthly_pi_usd: f64,
    /// Rent − operating − mortgage P&I.
    pub monthly_cash_flow_usd: f64,
    pub annual_cash_flow_usd: f64,
    /// ARV − refi loan: the equity you keep after pulling cash.
    pub equity_after_refi_usd: f64,
    /// Annual cash flow ÷ cash left in the deal. `None` when no cash is
    /// left (the return is infinite — the BRRRR ideal).
    pub cash_on_cash_pct: Option<f64>,
}

pub fn compute(i: &BrrrrInput) -> BrrrrResult {
    let total_cash_invested =
        i.purchase_price_usd + i.rehab_cost_usd + i.purchase_closing_usd;
    let refi_loan = i.after_repair_value_usd * (i.refi_ltv_pct / 100.0);
    let cash_out = refi_loan - i.refi_closing_usd;
    let cash_left = total_cash_invested - cash_out;

    let monthly_pi =
        crate::mortgage_amortization::monthly_payment(refi_loan, i.refi_apr_pct, i.refi_term_months);
    let monthly_cash_flow = i.monthly_rent_usd - i.monthly_operating_usd - monthly_pi;

    let cash_on_cash_pct = if cash_left > 0.0 {
        Some(monthly_cash_flow * 12.0 / cash_left * 100.0)
    } else {
        None
    };

    BrrrrResult {
        total_cash_invested_usd: total_cash_invested,
        refi_loan_usd: refi_loan,
        cash_out_usd: cash_out,
        cash_left_in_deal_usd: cash_left,
        all_cash_recovered: cash_left <= 0.0,
        monthly_pi_usd: monthly_pi,
        monthly_cash_flow_usd: monthly_cash_flow,
        annual_cash_flow_usd: monthly_cash_flow * 12.0,
        equity_after_refi_usd: i.after_repair_value_usd - refi_loan,
        cash_on_cash_pct,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Buy $100k + $30k rehab + $5k closing = $135k cash in.
    fn deal() -> BrrrrInput {
        BrrrrInput {
            purchase_price_usd: 100_000.0,
            rehab_cost_usd: 30_000.0,
            purchase_closing_usd: 5_000.0,
            after_repair_value_usd: 200_000.0,
            refi_ltv_pct: 75.0,
            refi_apr_pct: 7.0,
            refi_term_months: 360,
            refi_closing_usd: 4_000.0,
            monthly_rent_usd: 1_800.0,
            monthly_operating_usd: 600.0,
        }
    }

    #[test]
    fn successful_brrrr_recovers_all_cash() {
        // ARV $200k × 75% = $150k loan; net of $4k closing = $146k out
        // vs $135k in → all cash recovered, $11k pulled beyond basis.
        let r = compute(&deal());
        assert!((r.total_cash_invested_usd - 135_000.0).abs() < 1e-6);
        assert!((r.refi_loan_usd - 150_000.0).abs() < 1e-6);
        assert!((r.cash_out_usd - 146_000.0).abs() < 1e-6);
        assert!((r.cash_left_in_deal_usd - (-11_000.0)).abs() < 1e-6);
        assert!(r.all_cash_recovered);
        assert_eq!(r.cash_on_cash_pct, None); // infinite return
        assert!((r.equity_after_refi_usd - 50_000.0).abs() < 1e-6);
    }

    #[test]
    fn partial_brrrr_leaves_cash_and_has_finite_coc() {
        // Lower ARV → smaller loan → cash stays in the deal.
        let r = compute(&BrrrrInput { after_repair_value_usd: 160_000.0, ..deal() });
        // $160k × 75% = $120k loan; $116k out vs $135k in → $19k left.
        assert!((r.cash_left_in_deal_usd - 19_000.0).abs() < 1e-6);
        assert!(!r.all_cash_recovered);
        let coc = r.cash_on_cash_pct.expect("finite when cash is left");
        let expect = r.annual_cash_flow_usd / 19_000.0 * 100.0;
        assert!((coc - expect).abs() < 1e-9);
    }

    #[test]
    fn pi_and_cash_flow_use_shared_amortization() {
        let r = compute(&deal());
        let pi = crate::mortgage_amortization::monthly_payment(150_000.0, 7.0, 360);
        assert!((r.monthly_pi_usd - pi).abs() < 1e-9);
        assert!((r.monthly_cash_flow_usd - (1_800.0 - 600.0 - pi)).abs() < 1e-9);
        assert!((r.annual_cash_flow_usd - r.monthly_cash_flow_usd * 12.0).abs() < 1e-6);
    }

    #[test]
    fn equity_is_arv_minus_loan() {
        let r = compute(&BrrrrInput { refi_ltv_pct: 70.0, ..deal() });
        // $200k − $140k loan = $60k equity.
        assert!((r.equity_after_refi_usd - 60_000.0).abs() < 1e-6);
    }
}

File: frontend/js/views/brrrr.js
// BRRRR — Buy, Rehab, Rent, Refinance, Repeat. Nets the cash-out refi
// against total cash invested to show the cash left in the deal and the
// post-refi cash flow, via /calc/brrrr.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import * as enh from '../calc_enhance.js';

const VIEW = 'brrrr';
let lastReport = null;
let lastBody = null;

const FIELDS = [
    ['purchase_price_usd', 'Purchase price ($)', 100000],
    ['rehab_cost_usd', 'Rehab cost ($)', 30000],
    ['purchase_closing_usd', 'Acquisition closing ($)', 5000],
    ['after_repair_value_usd', 'After-repair value / ARV ($)', 200000],
    ['refi_ltv_pct', 'Refi LTV (%)', 75],
    ['refi_apr_pct', 'Refi APR (%)', 7],
    ['refi_term_months', 'Refi term (months)', 360],
    ['refi_closing_usd', 'Refi closing ($)', 4000],
    ['monthly_rent_usd', 'Monthly rent ($)', 1800],
    ['monthly_operating_usd', 'Monthly operating ($, ex-mortgage)', 600],
];
const INT_FIELDS = new Set(['refi_term_months']);

const money = (n) => (n < 0 ? '-$' : '$') + Math.abs(Number(n)).toLocaleString(undefined, { maximumFractionDigits: 0 });

export async function renderBrrrr(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.brrrr.h1.title">// BRRRR</span></h1>
        <p class="muted small" data-i18n="view.brrrr.hint.intro">
            Buy a distressed property cheap, rehab it to force appreciation, rent it, then
            refinance against the higher after-repair value and pull your cash back out to
            do it again. The deal succeeds when the cash-out refi recovers everything you
            put in — the cash left in the deal still earns rent, so the cash-on-cash return
            goes infinite.
        </p>
        <div class="chart-panel">
            <h2 data-i18n="view.brrrr.h2.inputs">The deal</h2>
            <form id="br-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.brrrr.label.${key}">${label}</span>
                        <input type="number" step="${INT_FIELDS.has(key) ? '1' : '0.01'}" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
                <button class="primary" type="submit" data-i18n="view.brrrr.btn.run">Run the deal</button>
            </form>
            <div id="br-tools" class="ce-toolbar"></div>
        </div>
        <div id="br-result"></div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#br-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        return body;
    };
    enh.mountToolbar(mount.querySelector('#br-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'brrrr.csv' });
    form.addEventListener('submit', async (e) => {
        e.preventDefault();
        const body = readBody();
        try {
            const r = await api.calcBrrrr(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.brrrr.toast.error'), { level: 'error' });
        }
    });
    form.dispatchEvent(new Event('submit'));
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['total_cash_invested_usd', r.total_cash_invested_usd],
        ['refi_loan_usd', r.refi_loan_usd],
        ['cash_out_usd', r.cash_out_usd],
        ['cash_left_in_deal_usd', r.cash_left_in_deal_usd],
        ['equity_after_refi_usd', r.equity_after_refi_usd],
        ['monthly_cash_flow_usd', r.monthly_cash_flow_usd],
    ];
}

function renderResult(mount, r) {
    const el = mount.querySelector('#br-result');
    const leftCls = Number(r.cash_left_in_deal_usd) <= 0 ? 'pos' : '';
    // Cash invested vs cash pulled at refi vs cash left in the deal.
    const chart = enh.svgBarChart([
        { label: 'Cash in', value: r.total_cash_invested_usd },
        { label: 'Cash out', value: r.cash_out_usd },
        { label: 'Left in', value: r.cash_left_in_deal_usd },
    ]);
    const cfCls = Number(r.monthly_cash_flow_usd) >= 0 ? 'pos' : 'neg';
    const coc = r.cash_on_cash_pct == null ? t('view.brrrr.infinite') : Number(r.cash_on_cash_pct).toFixed(1) + '%';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.brrrr.h2.result">The numbers</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.brrrr.card.invested">Total cash in</div>
                    <div class="value">${money(r.total_cash_invested_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.brrrr.card.loan">Refi loan</div>
                    <div class="value">${money(r.refi_loan_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.brrrr.card.cashout">Cash pulled at refi</div>
                    <div class="value pos">${money(r.cash_out_usd)}</div></div>
                <div class="card ${leftCls ? 'pos' : ''}"><div class="label" data-i18n="view.brrrr.card.left">Cash left in deal</div>
                    <div class="value ${leftCls}">${money(r.cash_left_in_deal_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.brrrr.card.equity">Equity after refi</div>
                    <div class="value">${money(r.equity_after_refi_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.brrrr.card.pi">Refi P&amp;I</div>
                    <div class="value">${money(r.monthly_pi_usd)}/mo</div></div>
                <div class="card"><div class="label" data-i18n="view.brrrr.card.cashflow">Monthly cash flow</div>
                    <div class="value ${cfCls}">${money(r.monthly_cash_flow_usd)}/mo</div></div>
                <div class="card"><div class="label" data-i18n="view.brrrr.card.coc">Cash-on-cash</div>
                    <div class="value ${r.cash_on_cash_pct == null ? 'pos' : ''}">${coc}</div></div>
            </div>
            ${chart}
            <p class="muted small">${r.all_cash_recovered
                ? `<span class="pos" data-i18n="view.brrrr.note.recovered">The refinance recovered all your cash — an infinite-return rental you can repeat.</span>`
                : `<span data-i18n="view.brrrr.note.partial">Some cash stays in the deal; the cash-on-cash return measures what it earns.</span>`}</p>
        </div>
    `;
    applyUiI18n(el);
}
